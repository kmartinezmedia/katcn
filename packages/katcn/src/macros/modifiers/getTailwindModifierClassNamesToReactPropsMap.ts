import { SyntaxKind } from 'ts-morph';
import type { ModifierProps } from '../../types';
import { getTwPrefixAndPropName } from '../getTwPrefixAndPropName';
import { katcnTypesSourceFile } from '../tsMorph';

/**
 * Used in tooling to convert plain tailwind classnames to katcn props
 *
 * This lookup map be used in conjuction with other ts-morph tooling
 * to convert `<div className="hover:bg-red-500" />` to
 * <Box _hover={{ bg: 'red-500' }} />
 *
 * Returns:
 *  { hover: '_hover', ...other modifiers }
 */
export function getTailwindModifierClassNamesToReactPropsMap() {
  const lookupMap = {} as Record<string, keyof ModifierProps>;
  const modifierPropsTypeDefinition =
    katcnTypesSourceFile
      .getInterface('ModifierProps')
      ?.getChildrenOfKind(SyntaxKind.PropertySignature) ?? [];

  for (const propertySignature of modifierPropsTypeDefinition) {
    if (!propertySignature) continue;
    const { name, twPrefix } = getTwPrefixAndPropName(propertySignature);
    lookupMap[twPrefix.replace(':', '')] = name as keyof ModifierProps;
  }
  return lookupMap;
}
