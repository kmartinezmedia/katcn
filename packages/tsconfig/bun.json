{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Bun",
  "extends": "./node.json",
  "compilerOptions": {
    "types": ["bun-types"]
  }
}
