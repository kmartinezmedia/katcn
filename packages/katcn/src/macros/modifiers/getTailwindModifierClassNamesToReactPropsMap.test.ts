import { describe, expect, it } from 'bun:test';
import { getTailwindModifierClassNamesToReactPropsMap } from './getTailwindModifierClassNamesToReactPropsMap';

describe('getTailwindModifierClassNamesToReactPropsMap', () => {
  it('Matches snapshot', () => {
    const value = getTailwindModifierClassNamesToReactPropsMap();
    expect(value).toMatchSnapshot();
  });
  it('Should return a mapping of tailwind modifier classnames (without colon separator) to katcn modifier props', () => {
    const value = getTailwindModifierClassNamesToReactPropsMap();
    expect(value).toBeDefined();
    expect(value).toMatchObject({
      hover: '_hover',
    });
  });
});
