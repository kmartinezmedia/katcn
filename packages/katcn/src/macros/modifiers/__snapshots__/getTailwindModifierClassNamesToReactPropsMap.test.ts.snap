// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getTailwindModifierClassNamesToReactPropsMap Matches snapshot 1`] = `
{
  "2xl": "_2xl",
  "active": "_active",
  "after": "_after",
  "aria-busy": "_ariaBusy",
  "aria-checked": "_ariaChecked",
  "aria-disabled": "_ariaDisabled",
  "aria-expanded": "_ariaExpanded",
  "aria-hidden": "_ariaHidden",
  "aria-pressed": "_ariaPressed",
  "aria-readonly": "_ariaReadonly",
  "aria-required": "_ariaRequired",
  "aria-selected": "_ariaSelected",
  "autofill": "_autofill",
  "backdrop": "_backdrop",
  "checked": "_checked",
  "cmdkGroup": "_cmdkGroup",
  "cmdkGroupHeading": "_cmdkGroupHeading",
  "cmdkGroupItems": "_cmdkGroupItems",
  "cmdkInput": "_cmdkInput",
  "cmdkInputSvg": "_cmdkInputSvg",
  "cmdkItem": "_cmdkItem",
  "cmdkItemSvg": "_cmdkItemSvg",
  "contrast-less": "_contrastLess",
  "contrast-more": "_contrastMore",
  "dark": "_dark",
  "data-[disabled=true]": "_dataDisabled",
  "default": "_default",
  "disabled": "_disabled",
  "empty": "_empty",
  "enabled": "_enabled",
  "even": "_even",
  "file": "_file",
  "first": "_first",
  "first-of-type": "_firstOfType",
  "focus": "_focus",
  "focus-visible": "_focusVisible",
  "focus-within": "_focusWithin",
  "forced-colors": "_forcedColors",
  "hover": "_hover",
  "in-range": "_inRange",
  "indeterminate": "_indeterminate",
  "invalid": "_invalid",
  "last": "_last",
  "last-of-type": "_lastOfType",
  "lg": "_lg",
  "md": "_md",
  "motion-reduce": "_motionReduce",
  "motion-safe": "_motionSafe",
  "odd": "_odd",
  "only": "_only",
  "only-of-type": "_onlyOfType",
  "out-of-range": "_outOfRange",
  "placeholder": "_placeholder",
  "placeholder-shown": "_placeholderShown",
  "read-only": "_readonly",
  "required": "_required",
  "selection": "_selection",
  "sm": "_sm",
  "target": "_target",
  "valid": "_valid",
  "visited": "_visited",
  "xl": "_xl",
}
`;
